### PR TITLE
fix: powershell redirection for filtered bang commands

### LINKED ISSUE #15913: Cannot create temp file for executing external commands over a range when shell is set to powershell
### Neovim version (nvim -v)

NVIM v0.5.1

### Vim (not Nvim) behaves the same?

Can't say. Do not have vim installed.

### Operating system/version

Windows 10 21H1

### Terminal name/version

Windows Terminal, Conhost

### $TERM environment variable

Nil

### Installation

Put the zip folder from release page and added the location in Windows path variable

### How to reproduce the issue

```vim
nvim -u NONE
:let &shell = 'pwsh'
:let &shellcmdflag = '-NoLogo -NoProfile -ExecutionPolicy RemoteSigned -Command [Console]::InputEncoding=[Console]::OutputEncoding=[System.Text.Encoding]::UTF8;'
:let &shellredir = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
:let &shellpipe = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
:set shellquote= shellxquote=
:e random-nonempty.file
:1,5!sort
```

### Expected behavior

The lines from 1 to 5 should be sorted with the external `sort` command (which exists in my path).

### Actual behavior

I get the following error.

```vim
E485: Can't read file C:\Users\ACER\AppData\Local\Temp\nvim6UJx04\7
```

I checked with Process Monitor and saw that neovim is failing to create the file/subdirectory `...\nvim6UJx04\7` when the `:h 'shell'` is set to `pwsh` (PowerShell Core). With shell set to cmd, it works alright. ~~Note that in my full config, I set up the shell according to `:h 'shell-powershell'` but the same issue persists.~~

### PR DESCRIPTION
**Pull request by 3N4N**
Revert #16271 and fix #15913.

Problem:
Since #16271, `make_filter_cmd` uses `Start-Process` cmdlet to execute the user
provided shell command for `:%!`. `Start-Process` requires the command to be
split into the shell command and its arguments. This was implemented in #19268
by parsing (splitting the user-provided command at the first space) which didn't
handle cases such as --
  - commands with escaped space in their filepath
  - quoted commands with space in their filepath

Solution: Use piping.

The total shell command formats (excluding noise of unimportant parameters):

1. Before #16271
    ```powershell
    pwsh -C "(shell_cmd) < tmp.in | 2>&1 Out-File -Encoding UTF8 <tmp.out>"
    # not how powershell commands work
    ```
2. Since #16271
    ```powershell
    pwsh -C "Start-Process shell_cmd -RedirectStandardInput <tmp.in> -RedirectStandardOutput <tmp.out>"
    # doesn't handle executable path with space in it
    # doesn't write error to <tmp.out>
    ```
3. This PR
    ```powershell
    pwsh -C "& { Get-Content <tmp.in> | & 'path\with space\to\shell_cmd.exe' arg1 arg2 } 2>&1 | Out-File -Encoding UTF8 <tmp.out>"
    # also works with forward slash in the filepath
    # also works with double quotes around shell command
    ```

After this PR, the user can use the following formats:
```
:%!c:\Program` Files\Git\usr\bin\sort.exe
:%!'c:\Program Files\Git\usr\bin\sort.exe'
:%!"c:\Program Files\Git\usr\bin\sort.exe"
:%!"c:\Program` Files\Git\usr\bin\sort.exe"
```

They can even chain different commands
```
:%!"c:\Program` Files\Git\usr\bin\sort.exe" | sort.exe -r
```

But if they want to call a stringed executable path, they have to provide the
Invoke-Command operator (&). In fact, the first stringed executable path also
needs this & operator, but this PR adds that behind the scene.
```
:%!"c:\Program` Files\Git\usr\bin\sort.exe" | sort.exe -r | & 'c:\Program Files\Git\usr\bin\sort.exe'
```

## What this PR solves

- Having to parse the user-provided bang ex-command (for splitting into shell
  cmd and its args).
- Removes a lot of human-unreadable `#ifdef` blocks.
- Accepting escaped spaces in executable path.
- Accepting quoted string of executable path.
- Redirects error and exception to tmp.out (exception for when `wrong_cmd.exe
  not found`)

## What this PR doesn't solve

- Handling wrongly escaped path to executable, which the user may pass because of cmdline tab-completion (see #18592).

## Edge cases
- [ ] If the user themself provides the `&` sign (means `call this.exe` in
  powershell) (I don't plan to handle it)
- [ ] Use `-Encoding utf8` parameter for `Get-Content`? (need to know about
  encoding, which I don't)
- [X] Doesn't write to tmp.out if shell command is not found (I have a fix:
  using anonymous function (`{wrong_cmd.exe}`))

## Changes other than `make_filter_cmd()` function

- Encoding (see [BOM-less UTF8][bomlessutf8]) for piping to external executables

[bomlessutf8]: https://github.com/PowerShell/PowerShell/issues/4681

### CODE REVIEW

File: src/nvim/ex_cmds.c [hunk 1]
```diff
@@ -1572,84 +1572,72 @@ char *make_filter_cmd(char *cmd, char *itmp, char *otmp)
...
-  len += is_fish_shell ?  sizeof("begin; " "; end") - 1
-                       :  is_pwsh ? sizeof("Start-Process ")
-                                  : sizeof("(" ")") - 1;
+  len += is_fish_shell ? sizeof("begin; " "; end") - 1
+                       : !is_pwsh ? sizeof("(" ")") - 1
+                                  : 0;
 
   if (itmp != NULL) {
-    len += is_pwsh  ? STRLEN(itmp) + sizeof(" -RedirectStandardInput ")
+    len += is_pwsh  ? STRLEN(itmp) + sizeof("Get-Content " " | & {  }") - 1
```
Comment (3N4N) [inline src/nvim/ex_cmds.c:1580]: (For myself) forgot to update the format here.

**Comment by zeertzjq** [comment]
Note: I may move the test to `ui/output_spec.lua` in #20361

**Comment by justinmk** [comment]
@3N4N thank you for your tenacity here. Do we need a doc update that mentions these points:

> The total shell command formats (excluding noise of unimportant parameters):...
>
> After this PR, the user can use the following formats:...
>
> But if they want to call a stringed executable path, they have to provide the Invoke-Command operator (&). In fact, the first stringed executable path also needs this & operator, but this PR adds that behind the scene.

?

**Comment by 3N4N** [comment]
@justinmk Other than the requirement of & before stringed executable path, users don't need to know about the others: it should just work any and every way they try.

We can document `&` like this below. But the example command is way too long. Can't think of a shorter command.
```
diff --git a/runtime/doc/various.txt b/runtime/doc/various.txt
index cae9c760308a..c2df8c914b25 100644
--- a/runtime/doc/various.txt
+++ b/runtime/doc/various.txt
@@ -259,6 +259,10 @@ g8			Print the hex values of the bytes used in the
 			or stderr, the streams are closed immediately. |E5677|
 			Use |jobstart()| instead. >
 				:call jobstart('foo', {'detach':1})
+
+			For powershell, chaining a stringed executable path
+			requires using the call operator (&).
+				:!Write-Output "1`n2" | & "C:\Windows\System32\sort.exe" /r
 <
 							*E34*
 			Any "!" in {cmd} is replaced with the previous
```

**Comment by justinmk** [comment]
Let's start with that. We want to write these things down, else they get lost in time.

**Comment by 3N4N** [comment]
> Let's start with that.

Do you mean push a new commit here, or a new PR?

**Comment by justinmk** [comment]
new PR

**Comment by sebshader** [comment]
one slight incompatibility I found with this is that on unix I seem to remember ```:.!``` not needing any input; here I get an error if the command (like Write-Output/echo) isn't able to take piped input. (I sometimes used to use uuidgen with this feature)

a workaround that does work with this pr is just to use ```? .;``` first like ```:.! ? .; echo "no piped input"``` instead, just thought I'd mention it.

**Pull request merged**